packages/clangd-contexts/src/clangd-config.ts

```diff
@@ -0,0 +1,102 @@
+export interface ClangdConfig {
+    CompileFlags: CompileFlags;
+    SourceFile: string;
+}
+
+/** Representation of the compile flags property of the `.clangd` configuration. */
+export interface CompileFlags {
+    Add?: string | string[];
+    Remove?: string | string[];
+    CompilationDatabase?: string;
+}
+
+/** The name of the project's clangd configuration file to which we apply a clangd context. */
+export const CONFIG_FILE_NAME = '.clangd';
+
+/**
+ * If a `.clangd` configuration file at the given path exists, load it into a {@link ClangdConfig}.
+ *
+ * @param configPath path to the clangd configuration file or its parent directory.
+ * @returns the loaded {@link ClangdConfig} or `undefined` if the configuration file does not exist
+ */
+export function load(configPath: string): ClangdConfig | undefined {
+    const configurationFile = toConfigPath(configPath);
+    if (!isFile(configurationFile)) {
+        return undefined;
+    }
+    const config = yaml.load(fs.readFileSync(configurationFile, 'utf8'));
+    if (config && typeof config === 'object') {
+        return createClangdConfig(configurationFile, config);
+    }
+    return undefined;
+}
+
+/**
+ * Create a new `ClangdConfig` object.
+ *
+ * @param configurationFile the configuration file from which the configuration was loaded or to which it will be written
+ * @param configurationData the configuration data loaded from the `configurationFile`, if any. This is generically typed
+ *  because it is usually loaded from YAML and may contain data used by clangd that is not managed by this API
+ * @returns a new clangd configuration object
+ */
+function createClangdConfig(configurationFile: string, configurationData?: object): ClangdConfig {
+    const result = { CompileFlags: {}, SourceFile: configurationFile };
+    if (configurationData) {
+        Object.assign(result, configurationData);
+    }
+    return result;
+}
+
+/**
+ * If a `.clangd` configuration file at the given path exists, load it into a {@link ClangdConfig},
+ * or else create a new empty {@link ClangdConfig}. A new empty file is not created; be sure not
+ * to {@link save} an empty configuration using the result.
+ *
+ * @param configPath path to the clangd configuration file or its parent directory
+ * @returns the loaded or new {@link ClangdConfig}
+ *
+ * @see {@link load}, {@link save}
+ */
+export function loadOrCreate(configPath: string): ClangdConfig {
+    const configurationFile = toConfigPath(configPath);
+    const config = load(configurationFile) ?? createClangdConfig(configurationFile);
+    return config;
+}
+
+/**
+ * Write the content of a given {@link ClangdConfig} to its {@link ClangdConfig.SourceFile source file}.
+ *
+ * @param config a clangd configuration to save
+ */
+export function save(config: ClangdConfig): void {
+    const sourceFile = config.SourceFile;
+    if (!sourceFile) {
+        console.error('Could not save config. No "SourceFile" has been defined.');
+    }
```

Comment:
`return;`? `throw`? Otherwise, the process proceeds under conditions that seem guaranteed to cause failure.

Reply:
I've added a `ClangConfigException` class to throw here and in the `load[OrCreate]` APIs. I've added a test for the exception thrown in loading an invalid YAML. But I've not updated the examples to do thorough exception handling. Let me know if that's a concern.